### ISSUE TITLE
Analysis/completion not updating after a change to a file

### ISSUE THREAD

**Issue by DanTup** [dart-lang/sdk#32766]
Found by @escamoteur, I can repro locally (only tried on Windows so far).

Repro code is here:
https://github.com/escamoteur/rx_command/tree/vs_code_not_updating

You can get as a zip here:
https://github.com/escamoteur/rx_command/archive/vs_code_not_updating.zip

Around line 55 of `lib/rx_command.dart` if you rename the `createSync3`/`createSync4` method to have another number on the end, and then switch to `test/rx_command_test.dart` the code that called it does not get an error as expected, and the completion still shows the old name for the method.

Here's the relevant parts of the instrumentation file.

[analyzer.instrum.1.txt](https://github.com/dart-lang/sdk/files/1876323/analyzer.instrum.1.txt)

@bwilkerson lmk if you can't repro with this code and require any additional logging/debugging.

**Comment by bwilkerson** [comment]
@scheglov

**Comment by scheglov** [comment]
Works for me on Mac.
If I change `createSync4` to `createSync45` by typing `5`, this causes a compilation error in `rx_command_test.dart`.

**Comment by bwilkerson** [comment]
Does the code that recognizes dependencies between files use the file paths? I'm wondering if this is an issue because of drive letter inconsistencies.

**Comment by DanTup** [comment]
Ugh, yes, works on my mac too. So seemingly another Windows issue. It could be the same issue we've seen elsewhere, where the VM has a different idea of drive letters for things it resolved paths for versus paths that came from the IDE.

**Comment by scheglov** [comment]
Yes, AnalysisDriver uses paths to get files. Dependencies between files (`FileState` objects) are recorded as `FileState` sets. `FileState` uses `uri` for hash code and equality.

**Comment by bwilkerson** [comment]
We need to fix `ResourceProvider`. Janice is running into a similar problem in dartdoc and thinks that the path package's `canonicalize`, `equals`, and `hash` functions should be able to solve the problem.

**Comment by scheglov** [comment]
`ResourceProvider` might be too late, there are other places where we use paths, or URIs derived from paths. I think it would be more practical to canonicalize all input paths, and canonicalize any URIs we get from directives.

**Comment by DanTup** [comment]
We tried to canonicalize paths in https://github.com/dart-lang/sdk/pull/32133 and it broke other things because those paths were used elsewhere (eg. as lookup keys). I'd hoped we could get some tests around these issues for Windows before attempting to fix again, but it's hard to create failing tests on Windows because the in-memory file system used in tests is case-sensitive so doesn't behave the same as real code on Windows.

**Comment by DanTup** [comment]
@escamoteur have you seen this at all recently? The original repro links are 404s now so I can't test with those. A lot has changed since this was opened so if it's not been seen recently and we don't have a repro, we should probably close this and file a new issue if it comes up again.

**Comment by escamoteur** [comment]
I haven't seen it lately. Only thing I constantly struggle is that the analyser takes sometimes for ages after for instance a `flutter packages upgrade` or a switch of branches to complete

**Comment by DanTup** [comment]
> I haven't seen it lately.

Thanks! In that case let's close this and open new issues if this is seen again.


>  Only thing I constantly struggle is that the analyser takes sometimes for ages after for instance a `flutter packages upgrade` or a switch of branches to complete

This can trigger a lot of analysis so it might take some time, but whether the times you see are normal or not is hard to say without more info. If it seems unreasonably long and you can collect some numbers, it might be worth filing in a new issue. If you can provide a repository and exact steps to reproduce (for example checking out a specific commit and changing to another specific commit, using a specific version of Dart/Flutter), that would definitely help.
